File: src/runtime/boxes/BbArbno.cs
// BbArbno.cs — ARBNO: zero-or-more greedy; zero-advance guard; β unwinds stack
// Mirrors src/runtime/boxes/bb_arbno.c
//
// Stack of (matched_spec, start_cursor) frames — each greedy iteration
// pushes one frame.  β pops the top frame (shortest match first on retry).
//
// α:  depth=0; push frame(zero-width, cursor);
//     loop: try body.α from current position
//       body_γ and cursor advanced → push new frame, loop
//       body_γ but cursor unchanged → zero-advance guard → stop, γ(top frame)
//       body_ω → γ(top frame)
// β:  if depth <= 0 → ω
//     depth--; restore cursor from frame; γ(frame.matched)

namespace Snobol4.Runtime.Boxes;

public sealed class BbArbno : IByrdBox
{
    private const int StackMax = 64;

    private readonly IByrdBox _body;

    private readonly Spec[] _matchedStack = new Spec[StackMax];
    private readonly int[]  _startStack   = new int[StackMax];
    private int              _depth;

    public BbArbno(IByrdBox body) { _body = body; }

    public Spec Alpha(MatchState ms)
    {
        _depth = 0;
        _matchedStack[0] = Spec.ZeroWidth(ms.Cursor);
        _startStack[0]   = ms.Cursor;

        while (true)
        {
            int startHere = ms.Cursor;
            var br = _body.Alpha(ms);

            if (br.IsFail)
            {
                // body_ω → stop, succeed with accumulated match
                return _matchedStack[_depth];        // ARBNO_γ
            }

            // body_γ — check zero-advance guard
            if (ms.Cursor == startHere)
            {
                // Zero advance: stop here (infinite-loop guard)
                return _matchedStack[_depth];        // ARBNO_γ_now
            }

            // Advance: push new frame
            if (_depth + 1 < StackMax)
            {
                _depth++;
                _matchedStack[_depth] = _matchedStack[_depth - 1].Cat(br);
                _startStack[_depth]   = ms.Cursor;
            }
            // else stack full: stop
            else
            {
                return _matchedStack[_depth];
            }
        }
    }

    public Spec Beta(MatchState ms)
    {
        if (_depth <= 0) return Spec.Fail;           // ARBNO_ω
        _depth--;
        ms.Cursor = _startStack[_depth];
        return _matchedStack[_depth];                // ARBNO_γ (shorter match)
    }
}

File: src/runtime/boxes/BbArbno.java
/**
 * BbArbno.java — ARBNO: zero-or-more greedy; zero-advance guard; β unwinds stack
 * Port of bb_arbno.c / bb_arbno.s
 *
 *   ARBNO_α:   depth=0; fr=stack[0]; fr.matched=spec(Σ+Δ,0); fr.start=Δ;
 *   ARBNO_try: br = body(α);  if empty → body_ω;  else → body_γ;
 *   ARBNO_β:   if (depth<=0)                       goto ARBNO_ω;
 *              depth--; fr=stack[depth]; Δ=fr.start; goto ARBNO_γ;
 *   body_γ:    fr=stack[depth];
 *              if (Δ==fr.start)                    goto ARBNO_γ_now;  (zero-advance guard)
 *              ARBNO=cat(fr.matched, br);
 *              if room: depth++; fr=stack[depth]; fr.matched=ARBNO; fr.start=Δ;
 *                       goto ARBNO_try;
 *   body_ω:    ARBNO=stack[depth].matched;         goto ARBNO_γ;
 *   ARBNO_γ_now: ARBNO=stack[depth].matched;       goto ARBNO_γ;
 *   ARBNO_γ:                                       return ARBNO;
 *   ARBNO_ω:                                       return spec_empty;
 */
class BbArbno extends BbBox {
    private static final int MAX_DEPTH = 64;

    private final BbBox body;

    /* per-frame saved state */
    private final int[] frameStart   = new int[MAX_DEPTH];
    private final int[] frameMatchSt = new int[MAX_DEPTH]; /* start of accumulated match */
    private final int[] frameMatchLn = new int[MAX_DEPTH]; /* length of accumulated match */
    private int depth;

    public BbArbno(MatchState ms, BbBox body) { super(ms); this.body=body; }

    @Override public Spec alpha() {
        // ARBNO_α
        depth = 0;
        frameMatchSt[0] = ms.delta;
        frameMatchLn[0] = 0;
        frameStart[0]   = ms.delta;
        return tryBody();
    }

    @Override public Spec beta() {
        // ARBNO_β: unwind one stack frame
        if (depth <= 0) return null;                                           // ARBNO_ω
        depth--;
        ms.delta = frameStart[depth];
        return new Spec(frameMatchSt[depth], frameMatchLn[depth]);             // ARBNO_γ
    }

    private Spec tryBody() {
        while (true) {
            Spec br = body.alpha();
            if (br == null) {
                // body_ω: stop looping, return current accumulated match
                return new Spec(frameMatchSt[depth], frameMatchLn[depth]);     // ARBNO_γ
            }
            // body_γ
            int savedStart = frameStart[depth];
            if (ms.delta == savedStart) {
                // zero-advance guard: don't push, return now
                return new Spec(frameMatchSt[depth], frameMatchLn[depth]);     // ARBNO_γ_now
            }
            int newMatchSt = frameMatchSt[depth];
            int newMatchLn = frameMatchLn[depth] + br.len;
            if (depth + 1 < MAX_DEPTH) {
                depth++;
                frameMatchSt[depth] = newMatchSt;
                frameMatchLn[depth] = newMatchLn;
                frameStart[depth]   = ms.delta;
                // ARBNO_try: loop
            } else {
                // stack full: accept current
                return new Spec(newMatchSt, newMatchLn);                       // ARBNO_γ
            }
        }
    }
}
